resctrl/utils.go

```diff
@@ -96,47 +106,53 @@ func prepareMonitoringGroup(containerName string, getContainerPids func() ([]str
 	}
 
 	if len(pids) == 0 {
-		return "", fmt.Errorf("couldn't obtain %q container pids, there is no pids in cgroup", containerName)
+		return "", fmt.Errorf("couldn't obtain %q container pids: there is no pids in cgroup", containerName)
 	}
 
-	path, err := findControlGroup(pids)
+	controlGroupPath, err := findGroup(rootResctrl, pids, true, false)
 	if err != nil {
-		return "", fmt.Errorf("couldn't find control group matching %q container: %v", containerName, err)
+		return "", fmt.Errorf("%q %q: %q", noControlGroupFoundError, containerName, err)
 	}
-
-	if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
-		containerName = containerName[1:]
+	if controlGroupPath == "" {
+		return "", fmt.Errorf("%q %q", noControlGroupFoundError, containerName)
 	}
 
-	properContainerName := strings.Replace(containerName, "/", "-", -1)
-	monGroupPath := filepath.Join(path, monitoringGroupDir, properContainerName)
-
-	// Create new mon_group if not exists.
-	err = os.MkdirAll(monGroupPath, os.ModePerm)
+	// Check if there is any monitoring group.
+	monGroupPath, err := findGroup(filepath.Join(controlGroupPath, monGroupsDirName), pids, false, true)
 	if err != nil {
-		return "", fmt.Errorf("couldn't create monitoring group directory for %q container: %w", containerName, err)
+		return "", fmt.Errorf("couldn't find monitoring group matching %q container: %v", containerName, err)
 	}
 
-	for _, pid := range pids {
-		threadFiles, err := ioutil.ReadDir(filepath.Join(processPath, pid, processTask))
-		if err != nil {
-			return "", err
+	// Prepare new one if not exists.
+	if monGroupPath == "" {
+		if len(containerName) >= minContainerNameLen && containerName[0] == containerPrefix {
+			containerName = containerName[1:]
 		}
 
-		processThreads, err := getAllProcessThreads(threadFiles)
+		properContainerName := fmt.Sprintf("%s-%s", monGroupPrefix, strings.Replace(containerName, "/", "-", -1))
+		monGroupPath = filepath.Join(controlGroupPath, monitoringGroupDir, properContainerName)
```

Comment:
is there a risk that e.g.
- kubepods/pod-x
- kubepods/pod-x/container-y
will be created and then you have an race condition, that only of those will contain pids (other will be stealed)

Reply:
8c734a03c1bc0e803b5b5cbbca03882ea0135bb5 will be sufficient for you?